File: frontend/src/api/equipment.ts
// 設備相關 API 呼叫封裝
import axios from 'axios'
import type { Equipment, EquipmentStatus, SpcDataPoint } from '../types/mes'

const BASE = '/api/v1'

// 後端回傳 snake_case，對應前端 camelCase 型別
function mapEquipment(raw: Record<string, unknown>): Equipment {
  return {
    id:           raw.id as number,
    name:         raw.name as string,
    type:         raw.type as Equipment['type'],
    status:       raw.status as EquipmentStatus,
    currentLotId: raw.current_lot_id as number | null,
    currentLot:   raw.current_lot as string | null,
    recipeName:   raw.recipe_name as string | null,
    utilization:  raw.utilization as number,
    temperature:  raw.temperature as number,
    pressure:     raw.pressure as number,
    ucl_temp:     raw.ucl_temp as number,
    lcl_temp:     raw.lcl_temp as number,
    ucl_pressure: raw.ucl_pressure as number,
    lcl_pressure: raw.lcl_pressure as number,
    isAlarm:      raw.is_alarm as boolean,
    updatedAt:    raw.updated_at as string,
  }
}

export async function getEquipments(): Promise<Equipment[]> {
  const { data } = await axios.get<{ data: Record<string, unknown>[] }>(`${BASE}/equipment`)
  return (data.data ?? []).map(mapEquipment)
}

export async function getEquipment(id: number): Promise<Equipment> {
  const { data } = await axios.get<{ data: Record<string, unknown> }>(`${BASE}/equipment/${id}`)
  return mapEquipment(data.data)
}

export async function updateEquipmentStatus(id: number, status: EquipmentStatus): Promise<void> {
  await axios.put(`${BASE}/equipment/${id}/status`, { status })
}

export async function holdEquipment(id: number): Promise<void> {
  await axios.post(`${BASE}/equipment/${id}/hold`)
}

// SPC 歷史資料（最近 N 筆）
export async function getSpcHistory(equipmentId: number, limit = 20): Promise<SpcDataPoint[]> {
  const { data } = await axios.get<{ data: Record<string, unknown>[] }>(
    `${BASE}/spc/${equipmentId}?limit=${limit}`
  )
  return (data.data ?? []).map(r => ({
    timestamp: r.timestamp as string,
    value:     r.value as number,
    ucl:       r.ucl as number,
    lcl:       r.lcl as number,
    isAlarm:   r.is_alarm as boolean,
  }))
}

File: backend/internal/model/equipment.go
package model

import "time"

type EquipmentStatus string

const (
	StatusIdle    EquipmentStatus = "IDLE"
	StatusRunning EquipmentStatus = "RUNNING"
	StatusDown    EquipmentStatus = "DOWN"
	StatusPM      EquipmentStatus = "PM"
)

// 合法的狀態轉換規則
var ValidTransitions = map[EquipmentStatus][]EquipmentStatus{
	StatusIdle:    {StatusRunning, StatusPM, StatusDown},
	StatusRunning: {StatusIdle, StatusDown},
	StatusDown:    {StatusPM},
	StatusPM:      {StatusIdle},
}

// IsValidTransition 驗證狀態轉換是否合法
func IsValidTransition(from, to EquipmentStatus) bool {
	targets, ok := ValidTransitions[from]
	if !ok {
		return false
	}
	for _, t := range targets {
		if t == to {
			return true
		}
	}
	return false
}

type Equipment struct {
	ID           int             `db:"id"             json:"id"`
	Name         string          `db:"name"           json:"name"`
	Type         string          `db:"type"           json:"type"`
	Status       EquipmentStatus `db:"status"         json:"status"`
	CurrentLotID *int            `db:"current_lot_id" json:"current_lot_id"`
	CurrentLot   *string         `db:"-"              json:"current_lot"`   // JOIN 欄位
	RecipeName   *string         `db:"-"              json:"recipe_name"`
	Utilization  float64         `db:"utilization"    json:"utilization"`
	Temperature  float64         `db:"temperature"    json:"temperature"`
	Pressure     float64         `db:"pressure"       json:"pressure"`
	UCLTemp      float64         `db:"ucl_temp"       json:"ucl_temp"`
	LCLTemp      float64         `db:"lcl_temp"       json:"lcl_temp"`
	UCLPressure  float64         `db:"ucl_pressure"   json:"ucl_pressure"`
	LCLPressure  float64         `db:"lcl_pressure"   json:"lcl_pressure"`
	IsAlarm      bool            `db:"is_alarm"       json:"is_alarm"`
	UpdatedAt    time.Time       `db:"updated_at"     json:"updated_at"`
}

// EquipmentGroup 群組折疊用
type EquipmentGroup struct {
	Type        string      `json:"type"`
	Equipments  []Equipment `json:"equipments"`
	AlarmCount  int         `json:"alarm_count"`
	Utilization float64     `json:"utilization"`
	StatusCount struct {
		Running int `json:"running"`
		Idle    int `json:"idle"`
		Down    int `json:"down"`
		PM      int `json:"pm"`
	} `json:"status_count"`
}
